1) список вхождений лемм в "Маленький принц"

In [1]:
# библиотека для работы с xml
# https://docs.python.org/3/library/xml.etree.elementtree.html
import xml.etree.ElementTree as ET

In [2]:
# прочитали и сделали дерево из большого файла (= "распарсили")
tree = ET.parse('/Users/tatianafrolova/Downloads/SynTagRus/2020_Corpus2_0/Malenkii_prints.tgt')

# находим главный тег (в этой библиотеке главный гет - root) и проверяем, как он называется
root = tree.getroot()
root.tag

'text'

In [3]:
# text - это правильно
# Какие есть теги внутри главного
# эта ячейка нужна, чтобы убедиться, что я правильн понимаю, как работает эта штука
# можно не запускать
for child in root:
     print(child.tag)

inf
body


In [4]:
# сохраним содержимое второго тега (body) в переменную body (по названию тега)
body = root[1]

In [5]:
# как называется первый тег внутри body
# если уже знать, можно не запускать
print(body[0].tag)

S


In [6]:
# посмотрим на теги внутри первого S
# эта ячейка нужна для несакольких следующих, а они нужны для того, чтобы разобраться с библиотекой и устройством файла
# если разбираться не нужно, можно не запускать
first_sentence = body[0]
for s_child in first_sentence:
  print(s_child.tag)

W
W
W
SEM


In [7]:
# внутри W тегов нет, есть атрибуты, они устроены как словарь
# для понимания устройства тега W
# можно не запускать
first_word = first_sentence[0]
print(first_word.attrib)

{'DOM': '_root', 'EXTRAFEAT': 'CAP ЗЕРО', 'FEAT': 'S ЕД МУЖ ИМ ОД', 'ID': '1', 'KSNAME': 'АНТУАН', 'LEMMA': 'АНТУАН'}


In [8]:
# все W первого предложения
# для понимания устройства тега W
# можно не запускать
for w_child in first_sentence:
  print(w_child.attrib)

{'DOM': '_root', 'EXTRAFEAT': 'CAP ЗЕРО', 'FEAT': 'S ЕД МУЖ ИМ ОД', 'ID': '1', 'KSNAME': 'АНТУАН', 'LEMMA': 'АНТУАН'}
{'DOM': '3', 'FEAT': 'PART', 'HYPOT': 'вспом.d0', 'ID': '2', 'KSNAME': 'ДЕ1', 'LEMMA': 'ДЕ', 'LINK': 'вспом'}
{'ACCTSTRING': 'Сент-Экзюпери', 'DOM': '1', 'EXTRAFEAT': 'CAP ЗЕРО', 'FEAT': 'S ЕД МУЖ ИМ ОД', 'HYPOT': 'аппоз.01', 'ID': '3', 'KSNAME': 'СЕНТ-ЭКЗЮПЕРИ', 'LEMMA': 'СЕНТ-ЭКЗЮПЕРИ', 'LINK': 'аппоз'}
{}


In [9]:
# в атрибуте LEMMA лежит имя леммы, оно мне и нужно
# пробуем вытащить значение по ключу
first_word.attrib['LEMMA']

'АНТУАН'

In [10]:
# в атрибуте LINK лежит имя связи
# пробуем вытащить значение по ключу
first_sentence[1].attrib['LINK']

'вспом'

если атрибута LINK в этом теге нет, то ничего не получится
пробуем вытащить значение по ключу
такой запрос выдаст ошибку
first_sentence[0].attrib['LINK']

В описании ET есть такой способ, он сюда вроде бы подходит:

get(key, default=None)
Gets the element attribute named key.

Returns the attribute value, or default if the attribute was not found.

In [11]:
# внимание скобки! здесь круглые - это функция, а не квадратные как после attrib без get - там типа залезание в словарь
first_sentence[0].attrib.get('LINK','root')

'root'

In [12]:
# заводим список, в который будем сохранять леммы и такой же для связей
prince_lemmas = []
prince_links = []

In [13]:
# проверяю, что так сработает цикл с i
for i in range(0, 4):
  print(first_sentence[i].tag)

W
W
W
SEM


In [14]:
# теперь нужно сделать цикл, который извлекает все леммы из слов одного предложения и складывает в список
# список каждый раз делаем пустой перед работой, чтобы не было лишних повторов записи
prince_lemmas = []
for i in range(0, 4):
  if first_sentence[i].tag == 'W':
    prince_lemmas.append(first_sentence[i].attrib['LEMMA'])

print(prince_lemmas)

['АНТУАН', 'ДЕ', 'СЕНТ-ЭКЗЮПЕРИ']


In [15]:
prince_links = []
for i in range(0, 4):
  if first_sentence[i].tag == 'W':
    prince_links.append(first_sentence[i].attrib.get('LINK','root'))

print(prince_links)


['root', 'вспом', 'аппоз']


In [16]:
# я задавала range [0,4], зная длину предложения
# как задать range для любого предложения?
prince_lemmas = []
for i in range(len(first_sentence)):
  if first_sentence[i].tag == 'W':
    prince_lemmas.append(first_sentence[i].attrib['LEMMA'])

print(prince_lemmas)

['АНТУАН', 'ДЕ', 'СЕНТ-ЭКЗЮПЕРИ']


In [17]:
# теперь нужно сделать цикл, который извлекает все леммы из всех слов всех предложений
# бывают ли пустые леммы? если есть, есть ли проблема записать их в список?
# попробуем  засунуть цикл из предыдущей ячейки в функцию
def find_lemmas(j):
  for i in range(len(body[j])):
    if body[j][i].tag == 'W':
      prince_lemmas.append(body[j][i].attrib['LEMMA'])

In [18]:
# проверим, как она работает на предложении 3
prince_lemmas = []
find_lemmas(2)
print(prince_lemmas)

['НОРА', 'ГАЛЬ', 'ПЕРЕВОД']


In [19]:
# попробуем применить эту функцию ко всем предложениям текста
prince_lemmas = []
for j in range(len(body)):
   find_lemmas(j)
  
print(len(prince_lemmas))

12257


In [20]:
# аналогичная функция для связей
def find_links(j):
  for i in range(len(body[j])):
    if body[j][i].tag == 'W':
      prince_links.append(body[j][i].attrib.get('LINK','root'))

In [21]:
# проверяем на предложении 3
prince_links = []
find_links(2)
print(prince_links)

['root', 'аппоз', 'сочин']


In [22]:
# попробуем применить эту функцию ко всем предложениям текста
prince_links = []
for j in range(len(body)):
   find_links(j)
  
print(len(prince_links))

12257


In [23]:
prince_links[0:10]

['root',
 'вспом',
 'аппоз',
 'опред',
 'root',
 'root',
 'аппоз',
 'сочин',
 'root',
 'аппоз']

In [24]:
# всё верно, число совпадает - у каждого слова либо имя входящей связи, либо указание на то, что это вершина
# теперь нужно сохранить эти списки в файлы, чтобы можно было выложить их и проверять работу без текстов
import os

with open("файл.txt", "w") as file:
    print(*список, file=file)

In [25]:
with open('prince_lemmas.txt', 'w') as file: 
    print(prince_lemmas, file = file)
    
with open('prince_links.txt', 'w') as file: 
    print(prince_links, file = file)

2) список имен файлов

In [26]:
#попытки сделать список путей через модуль os не увенчались успехом,
# вот здесь написано, как это сделать:
# https://realpython.com/get-all-files-in-directory-python/#getting-a-list-of-all-files-and-folders-in-a-directory-in-python
# при этом очень удачно пригодилось, что не получилось поменять расширение файлов с tgt на txt
import pathlib 
SynTagRus_path = pathlib.Path("/Users/tatianafrolova/Downloads/SynTagRus")
file_list = list(SynTagRus_path.rglob("*.tgt")) 
# если просто glob без r, то достает файлы с нужным расширением только в указанной папке 
#(а в указанной папке еще вложенные папки, а файлы ниже)
# r - это типа рекурсия, поэтому эта штука полезет внутрь всех папок внутри указанной
print(len(file_list))
file_list[1:10]

1304


[PosixPath('/Users/tatianafrolova/Downloads/SynTagRus/2013/Sverkhnovaya_ekonomika.tgt'),
 PosixPath('/Users/tatianafrolova/Downloads/SynTagRus/2013/Podarok_dlya_vsekh.tgt'),
 PosixPath('/Users/tatianafrolova/Downloads/SynTagRus/2013/Povedenie.tgt'),
 PosixPath('/Users/tatianafrolova/Downloads/SynTagRus/2013/Transplantatsiya_organov.tgt'),
 PosixPath('/Users/tatianafrolova/Downloads/SynTagRus/2013/Uchebnik_istorii.tgt'),
 PosixPath('/Users/tatianafrolova/Downloads/SynTagRus/2013/Sumerki_rossiiskoi_korruptsii.tgt'),
 PosixPath('/Users/tatianafrolova/Downloads/SynTagRus/2013/Poslednyaya_taina.tgt'),
 PosixPath('/Users/tatianafrolova/Downloads/SynTagRus/2013/A_v_chem_vinovat_Kipr.tgt'),
 PosixPath('/Users/tatianafrolova/Downloads/SynTagRus/2013/Doroga_v_nikuda.tgt')]

1304 это похоже на правду
Чтобы получить название без пути находим на 
https://docs.python.org/3/library/pathlib.html
вот такой способ:
PurePath.stem
The final path component, without its suffix:
    PurePosixPath('my/library.tar.gz').stem
'library.tar'
>>> PurePosixPath('my/library.tar').stem
'library'
>>> PurePosixPath('my/library').stem
'library'

In [27]:
# проверяем название
file_list[1].stem

'Sverkhnovaya_ekonomika'

In [28]:
# делаю список имен файлов без пути - понадобится для таблицы tfidf
file_names_list = []
for f in file_list:
    file_names_list.append(f.stem)
print(file_names_list[1:10])
len(file_names_list)

['Sverkhnovaya_ekonomika', 'Podarok_dlya_vsekh', 'Povedenie', 'Transplantatsiya_organov', 'Uchebnik_istorii', 'Sumerki_rossiiskoi_korruptsii', 'Poslednyaya_taina', 'A_v_chem_vinovat_Kipr', 'Doroga_v_nikuda']


1304

In [29]:
# сохраняем в файл
with open('corpus_files_names.txt', 'w') as file: 
    print(file_names_list, file = file)

3) список вхождений лемм в весь корпус

На память о том, как не подошла вроде бы работающая функция:
ужасно долго работает и приходится всё прерывать, если парсинг внутри цикла.
Нужно открыть файл один раз для каждого цикла

In [30]:
#  делаем функцию от двух переменных - каждый файл (то есть его BODY), каждый тег S
def find_all_lemmas(body, j):
    for i in range(len(body[j])):
        if body[j][i].tag == 'W':
            corpus_lemmas.append(body[j][i].attrib['LEMMA'])

# проверяем на маленьком кусочке одного файла
corpus_lemmas = []    
find_all_lemmas(ET.parse('/Users/tatianafrolova/Downloads/SynTagRus/2020_Corpus2_0/Malenkii_prints.tgt').getroot()[1], 2)
print(corpus_lemmas)

['НОРА', 'ГАЛЬ', 'ПЕРЕВОД']


In [31]:
# проходим функцией по всем файлам корпуса
corpus_lemmas = [] 
for k in range(len(file_list)):
    body = ET.parse(file_list[k]).getroot()[1]
    for j in range(len(body)):
        find_all_lemmas(body, j)
print(len(corpus_lemmas))

1515986


Полтора миллиона словоупотреблений. Соответствует информации с https://ruscorpora.ru

In [32]:
#  та же функция для синт онтошений
def find_all_links(body, j):
    for i in range(len(body[j])):
        if body[j][i].tag == 'W':
            corpus_links.append(body[j][i].attrib.get('LINK','root'))

corpus_links = []    
find_all_links(ET.parse('/Users/tatianafrolova/Downloads/SynTagRus/2020_Corpus2_0/Malenkii_prints.tgt').getroot()[1], 2)
print(corpus_links)

['root', 'аппоз', 'сочин']


In [33]:
# проходим функцией по всем файлам корпуса
corpus_links = [] 
for k in range(len(file_list)):
    body = ET.parse(file_list[k]).getroot()[1]
    for j in range(len(body)):
        find_all_links(body, j)
print(len(corpus_links))

1515986


In [34]:
# Сохраняем в файл
with open('corpus_lemmas.txt', 'w') as file: 
    print(corpus_lemmas, file = file)
    
with open('corpus_links.txt', 'w') as file: 
    print(corpus_links, file = file)

4) список списков вхождений лемм в каждый текст корпуса
для подсчета tfidf нужен список с именами файлов и список списков лемм, соответсвующих каждому файлу

In [35]:
# функция для списка лемм одного предложения 
def find_sentence_lemmas(body, j):
    lemmas = []    
    for i in range(len(body[j])):
        if body[j][i].tag == 'W':
            lemmas.append(body[j][i].attrib['LEMMA'])
    return lemmas

print((find_sentence_lemmas(ET.parse('/Users/tatianafrolova/Downloads/SynTagRus/2020_Corpus2_0/Malenkii_prints.tgt').getroot()[1], 2)))
print((find_sentence_lemmas(ET.parse('/Users/tatianafrolova/Downloads/SynTagRus/2020_Corpus2_0/Malenkii_prints.tgt').getroot()[1], 3)))

['НОРА', 'ГАЛЬ', 'ПЕРЕВОД']
['ЛЕОН', 'ВЕРТ']


In [ ]:
# Объединить списки для предложения в список для текста
# append(А) - добавит список А в качестве элемента, extend - расширит обрабатываемый список на элементы списка А
def find_text_lemmas(body):
    text_lemmas = []
    for j in range(len(body)):
       text_lemmas.extend(find_sentence_lemmas(body, j))
    return text_lemmas

print((find_text_lemmas(ET.parse('/Users/tatianafrolova/Downloads/SynTagRus/2020_Corpus2_0/Malenkii_prints.tgt').getroot()[1])))

['АНТУАН', 'ДЕ', 'СЕНТ-ЭКЗЮПЕРИ', 'МАЛЕНЬКИЙ', 'ПРИНЦ', 'НОРА', 'ГАЛЬ', 'ПЕРЕВОД', 'ЛЕОН', 'ВЕРТ', 'ПРОСИТЬ', 'РЕБЕНОК', 'ПРОЩАТЬ', 'Я', 'ЗА', 'ТО', 'ЧТО', 'Я', 'ПОСВЯЩАТЬ', 'ЭТОТ', 'КНИЖКА', 'ВЗРОСЛЫЙ', 'СКАЗАТЬ', 'В', 'ОПРАВДАНИЕ', 'ЭТОТ', 'ВЗРОСЛЫЙ', 'МОЙ', 'САМЫЙ', 'ЛУЧШИЙ', 'ДРУГ', 'И', 'ЕЩЕ', 'ОН', 'ПОНИМАТЬ', 'ВСЕ', 'НА', 'СВЕТ', 'ДАЖЕ', 'ДЕТСКИЙ', 'КНИЖКА', 'И', 'НАКОНЕЦ', 'ОН', 'ЖИТЬ', 'В', 'ФРАНЦИЯ', 'А', 'ТАМ', 'СЕЙЧАС', 'ГОЛОДНЫЙ', 'И', 'ХОЛОДНЫЙ', 'И', 'ОН', 'ОЧЕНЬ', 'НУЖДАТЬСЯ', 'В', 'УТЕШЕНИЕ', 'ЕСЛИ', 'ЖЕ', 'ВЕСЬ', 'ЭТО', 'Я', 'НЕ', 'ОПРАВДЫВАТЬ', 'Я', 'ПОСВЯЩАТЬ', 'ЭТОТ', 'КНИЖКА', 'ТОТ', 'МАЛЬЧИК', 'КАКОЙ', 'БЫТЬ', 'КОГДА-ТО', 'МОЙ', 'ВЗРОСЛЫЙ', 'ДРУГ', 'ВЕДЬ', 'ВЕСЬ', 'ВЗРОСЛЫЙ', 'СНАЧАЛА', 'БЫТЬ', 'РЕБЕНОК', 'ТОЛЬКО', 'МАЛО', 'КТО', 'ИЗ', 'ОНИ', 'О', 'ЭТО', 'ПОМНИТЬ', 'ИТАК', 'Я', 'ИСПРАВЛЯТЬ', 'ПОСВЯЩЕНИЕ', 'ЛЕОН', 'ВЕРТ', 'КОГДА', 'ОН', 'БЫТЬ', 'МАЛЕНЬКИЙ', 'I', 'КОГДА', 'Я', 'БЫТЬ', 'ШЕСТЬ', 'ГОД', 'В', 'КНИГА', 'ПОД', 'НАЗВАНИЕ', 'ПРАВДИВЫЙ', 'ИСТОРИЯ', 'ГДЕ', '

In [37]:
# применим к корпусу
# здесь нужно делать не extend, а append (добавит элемент) и join (сделает из списка строку)
# потому что именно с таким объектом работает TfidfVectorizer
texts_in_corpus_lemmas = []
for f in file_list:
    texts_in_corpus_lemmas.append(' '.join(find_text_lemmas(ET.parse(f).getroot()[1])))
print(len(texts_in_corpus_lemmas))

1304


In [38]:
#сохраним в файл
with open('texts_in_corpus_lemmas.txt', 'w') as file: 
    print(texts_in_corpus_lemmas, file = file)

In [39]:
#повторяем для отношений
# для предложений
def find_sentence_links(body, j):
    links = []    
    for i in range(len(body[j])):
        if body[j][i].tag == 'W':
            links.append(body[j][i].attrib.get('LINK','root'))
    return links

print((find_sentence_links(ET.parse('/Users/tatianafrolova/Downloads/SynTagRus/2020_Corpus2_0/Malenkii_prints.tgt').getroot()[1], 2)))
print((find_sentence_links(ET.parse('/Users/tatianafrolova/Downloads/SynTagRus/2020_Corpus2_0/Malenkii_prints.tgt').getroot()[1], 3)))

['root', 'аппоз', 'сочин']
['root', 'аппоз']


In [ ]:
# Объединить списки для предложения в список для текста
# append(А) - добавит список А в качестве элемента, extend - расширит обрабатываемый список на элементы списка А
def find_text_links(body):
    text_links = []
    for j in range(len(body)):
       text_links.extend(find_sentence_links(body, j))
    return text_links

print((find_text_links(ET.parse('/Users/tatianafrolova/Downloads/SynTagRus/2020_Corpus2_0/Malenkii_prints.tgt').getroot()[1])))

['root', 'вспом', 'аппоз', 'опред', 'root', 'root', 'аппоз', 'сочин', 'root', 'аппоз', 'root', '1-компл', '2-компл', '1-компл', '3-компл', 'предл', 'эксплет', 'предик', 'подч-союзн', 'опред', '1-компл', '2-компл', 'root', 'обст', 'предл', 'опред', 'предик', 'опред', 'опред', 'опред', '1-компл', 'root', 'соч-союзн', 'предик', 'разъяснит', '1-компл', 'атриб', 'предл', 'огранич', 'опред', 'сочин', 'root', 'вводн', 'предик', 'соч-союзн', '1-компл', 'предл', 'сент-соч', 'обст', 'обст', 'соч-союзн', 'сочин', 'соч-союзн', 'root', 'предик', 'огранич', 'соч-союзн', '1-компл', 'предл', 'обст', 'огранич', 'опред', 'предик', '1-компл', 'огранич', 'подч-союзн', 'предик', 'root', 'опред', '1-компл', 'опред', '2-компл', 'присвяз', 'релят', 'обст', 'опред', 'опред', 'предик', 'огранич', 'опред', 'предик', 'обст', 'root', 'присвяз', 'сент-соч', 'вспом', 'предик', 'атриб', 'предл', '2-компл', 'предл', 'соч-союзн', 'вводн', 'предик', 'root', '1-компл', 'root', 'аппоз', 'обст', 'предик', 'релят', 'присвяз

In [41]:
# применим к корпусу
texts_in_corpus_links = []
for f in file_list:
    texts_in_corpus_links.append(' '.join(find_text_links(ET.parse(f).getroot()[1])))
print(len(texts_in_corpus_links))

1304


In [42]:
texts_in_corpus_links[0]

'обст предл сочин соч-союзн опред предл обст опред предл root предик количест опред квазиагент 1-компл релят предик 2-компл сочин соч-союзн опред предик 1-компл аппоз аппоз аппоз аппоз root обст опред предл огранич опред 1-компл 1-компл опред предл квазиагент предик релят 1-компл 3-компл предл предик вводн root композ опред 1-компл сочин 2-компл опред опред соч-союзн предик релят огранич 1-компл 1-компл 4-компл опред предл обст количест предл квазиагент огранич опред предик root 1-компл 2-компл сочин соч-союзн 1-компл предл вводн опред предик обст опред предл предик аппоз огранич root опред 1-компл 3-компл опред сочин соч-союзн предл предик огранич root 1-компл предл квазиагент 1-компл 1-компл предл опред 1-компл присвяз квазиагент изъясн предик сочин соч-союзн сочин опред сочин соч-союзн квазиагент атриб предл опред опред 1-компл предик root 1-компл предик огранич подч-союзн обст опред предл 1-компл квазиагент 1-компл огранич root обст предл опред сочин соч-союзн 1-компл сочин 1-компл

In [43]:
#сохраним в файл
with open('texts_in_corpus_links.txt', 'w') as file: 
    print(texts_in_corpus_links, file = file)